File: frontend/src/api/buildable.ts
function normaliseBaseUrl(value: string | undefined | null): string {
  if (!value) {
    return '/'
  }
  const trimmed = value.trim()
  return trimmed === '' ? '/' : trimmed
}

const API_PREFIX = 'api/v1/screen/buildable'
const rawApiBaseUrl = import.meta.env.VITE_API_BASE_URL
const apiBaseUrl = normaliseBaseUrl(rawApiBaseUrl)

export type RuleProvenance = {
  rule_id: number
  clause_ref?: string | null
  document_id?: number | null
  pages?: number[] | null
  seed_tag?: string | null
  source?: string | null
}

export type RuleItem = {
  id: number
  authority: string
  parameter_key: string
  operator: string
  value: string
  unit?: string | null
  provenance: RuleProvenance
}

export type ZoneSource = {
  kind: 'parcel' | 'geometry' | 'unknown'
  layer_name?: string | null
  jurisdiction?: string | null
  parcel_ref?: string | null
  parcel_source?: string | null
  note?: string | null
}

export type BuildableResponse = {
  input_kind: 'address' | 'geometry'
  zone_code: string | null
  overlays: string[]
  advisory_hints?: string[] | null
  metrics: {
    gfa_cap_m2: number
    floors_max: number
    footprint_m2: number
    nsa_est_m2: number
  }
  zone_source: ZoneSource
  rules: RuleItem[]
}

export interface BuildableRequest {
  address: string
  typFloorToFloorM: number
  efficiencyRatio: number
}

export interface BuildableMetrics {
  gfaCapM2: number
  floorsMax: number
  footprintM2: number
  nsaEstM2: number
}

export interface BuildableRuleProvenance {
  ruleId: number
  clauseRef?: string
  documentId?: number
  pages?: number[]
  seedTag?: string
}

export interface BuildableRule {
  id: number
  authority: string
  parameterKey: string
  operator: string
  value: string
  unit?: string
  provenance: BuildableRuleProvenance
}

export interface ZoneSourceInfo {
  kind: 'parcel' | 'geometry' | 'unknown'
  layerName?: string
  jurisdiction?: string
  parcelRef?: string
  parcelSource?: string
  note?: string
}

export interface BuildableSummary {
  inputKind: 'address' | 'geometry'
  zoneCode: string | null
  overlays: string[]
  advisoryHints: string[]
  metrics: BuildableMetrics
  zoneSource: ZoneSourceInfo
  rules: BuildableRule[]
}

export interface BuildableTransportOptions {
  signal?: AbortSignal
}

export type BuildableTransport = (
  baseUrl: string,
  body: BuildableRequest,
  options?: BuildableTransportOptions,
) => Promise<BuildableResponse>

export interface BuildableRequestOptions extends BuildableTransportOptions {
  transport?: BuildableTransport
}

function buildUrl(path: string, base: string = apiBaseUrl) {
  if (/^https?:/i.test(path)) {
    return path
  }
  const trimmed = path.startsWith('/') ? path.slice(1) : path
  const root = normaliseBaseUrl(base)
  if (/^https?:/i.test(root)) {
    return new URL(trimmed, root.endsWith('/') ? root : `${root}/`).toString()
  }
  const normalisedRoot = root.endsWith('/') ? root : `${root}/`
  return `${normalisedRoot}${trimmed}`
}

function mapRule(rule: RuleItem): BuildableRule {
  return {
    id: rule.id,
    authority: rule.authority,
    parameterKey: rule.parameter_key,
    operator: rule.operator,
    value: rule.value,
    unit: rule.unit ?? undefined,
    provenance: {
      ruleId: rule.provenance.rule_id,
      clauseRef: rule.provenance.clause_ref ?? undefined,
      documentId: rule.provenance.document_id ?? undefined,
      pages: rule.provenance.pages ?? undefined,
      seedTag: rule.provenance.seed_tag ?? undefined,
    },
  }
}

function mapResponse(payload: BuildableResponse): BuildableSummary {
  const advisoryHints = Array.isArray(payload.advisory_hints)
    ? payload.advisory_hints.filter((hint): hint is string => typeof hint === 'string')
    : []

  return {
    inputKind: payload.input_kind,
    zoneCode: payload.zone_code,
    overlays: [...payload.overlays],
    advisoryHints: [...advisoryHints],
    metrics: {
      gfaCapM2: payload.metrics.gfa_cap_m2,
      floorsMax: payload.metrics.floors_max,
      footprintM2: payload.metrics.footprint_m2,
      nsaEstM2: payload.metrics.nsa_est_m2,
    },
    zoneSource: {
      kind: payload.zone_source.kind,
      layerName: payload.zone_source.layer_name ?? undefined,
      jurisdiction: payload.zone_source.jurisdiction ?? undefined,
      parcelRef: payload.zone_source.parcel_ref ?? undefined,
      parcelSource: payload.zone_source.parcel_source ?? undefined,
      note: payload.zone_source.note ?? undefined,
    },
    rules: payload.rules.map(mapRule),
  }
}

export async function postBuildable(
  baseUrl: string,
  body: BuildableRequest,
  options: BuildableTransportOptions = {},
): Promise<BuildableResponse> {
  const response = await fetch(buildUrl(API_PREFIX, baseUrl), {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      address: body.address,
      typ_floor_to_floor_m: body.typFloorToFloorM,
      efficiency_ratio: body.efficiencyRatio,
    }),
    signal: options.signal,
  })

  const contentType = response.headers.get('content-type') ?? ''

  if (!response.ok) {
    if (contentType.includes('application/json')) {
      const payload = (await response.json()) as { detail?: unknown; error?: unknown }
      const detail = typeof payload.detail === 'string' ? payload.detail : undefined
      const error = typeof payload.error === 'string' ? payload.error : undefined
      throw new Error(
        detail ?? error ?? `Request to /${API_PREFIX} failed with ${response.status}`,
      )
    }
    const message = await response.text()
    throw new Error(message || `Request to /${API_PREFIX} failed with ${response.status}`)
  }

  if (!contentType.includes('application/json')) {
    throw new Error(`Expected JSON response from /${API_PREFIX}`)
  }

  return (await response.json()) as BuildableResponse
}

export async function fetchBuildable(
  request: BuildableRequest,
  options: BuildableRequestOptions = {},
): Promise<BuildableSummary> {
  const { transport = postBuildable, signal } = options
  const payload = await transport(apiBaseUrl, request, { signal })
  return mapResponse(payload)
}

File: backend/app/schemas/buildable.py
"""Pydantic schemas for buildable screening."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    computed_field,
    field_validator,
    model_validator,
)

from app.core.config import settings

_BUILDABLE_DEFAULTS_EXAMPLE = {
    "plot_ratio": 3.5,
    "site_area_m2": 1000.0,
    "site_coverage": 0.45,
    "floor_height_m": 4.0,
    "efficiency_factor": 0.82,
}


_BUILDABLE_REQUEST_EXAMPLE = {
    "address": "string",
    "geometry": {"string": None},
    "project_type": "string",
    "defaults": _BUILDABLE_DEFAULTS_EXAMPLE,
    "typ_floor_to_floor_m": 4.0,
    "efficiency_ratio": 0.82,
}


_BUILDABLE_RESPONSE_EXAMPLE = {
    "input_kind": "address",
    "zone_code": "string",
    "overlays": ["string"],
    "advisory_hints": ["string"],
    "metrics": {
        "gfa_cap_m2": 0,
        "floors_max": 0,
        "footprint_m2": 0,
        "nsa_est_m2": 0,
    },
    "zone_source": {
        "kind": "parcel",
        "layer_name": "string",
        "jurisdiction": "string",
        "parcel_ref": "string",
        "parcel_source": "string",
        "note": "string",
    },
    "rules": [
        {
            "id": 0,
            "authority": "string",
            "parameter_key": "string",
            "operator": "string",
            "value": "string",
            "unit": "string",
            "provenance": {
                "rule_id": 0,
                "clause_ref": "string",
                "document_id": 0,
                "pages": [0],
                "seed_tag": "string",
            },
        }
    ],
}

BUILDABLE_REQUEST_EXAMPLE = _BUILDABLE_REQUEST_EXAMPLE
BUILDABLE_RESPONSE_EXAMPLE = _BUILDABLE_RESPONSE_EXAMPLE


class BuildableDefaults(BaseModel):
    """Default assumptions for buildable calculations."""

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={"example": _BUILDABLE_DEFAULTS_EXAMPLE},
    )

    plot_ratio: float = Field(default=3.5, gt=0, alias="plotRatio")
    site_area_m2: float = Field(default=1000.0, gt=0, alias="siteAreaM2")
    site_coverage: float = Field(default=0.45, gt=0, alias="siteCoverage")
    floor_height_m: float = Field(
        default=settings.BUILDABLE_TYP_FLOOR_TO_FLOOR_M,
        gt=0,
        alias="floorHeightM",
    )
    efficiency_factor: float = Field(
        default=settings.BUILDABLE_EFFICIENCY_RATIO,
        gt=0,
        alias="efficiencyFactor",
    )

    @field_validator("site_coverage")
    @classmethod
    def _normalise_site_coverage(cls, value: float) -> float:
        """Allow site coverage to be provided as a fraction or percentage."""

        if value > 1:
            value = value / 100.0
        return max(0.0, min(value, 1.0))

    @field_validator("efficiency_factor")
    @classmethod
    def _limit_efficiency(cls, value: float) -> float:
        """Ensure efficiency factor stays within a 0-1 range."""

        return max(0.0, min(value, 1.0))


class BuildableRequest(BaseModel):
    """Request payload for buildable screening."""

    model_config = ConfigDict(
        populate_by_name=True,
        json_schema_extra={"example": _BUILDABLE_REQUEST_EXAMPLE},
    )

    address: str | None = None
    geometry: dict[str, Any] | None = None
    project_type: str | None = None
    defaults: BuildableDefaults = Field(
        default_factory=BuildableDefaults, alias="defaults"
    )
    typ_floor_to_floor_m: float | None = Field(
        default=settings.BUILDABLE_TYP_FLOOR_TO_FLOOR_M,
        gt=0,
        alias="typFloorToFloorM",
    )
    efficiency_ratio: float | None = Field(
        default=settings.BUILDABLE_EFFICIENCY_RATIO,
        gt=0,
        alias="efficiencyRatio",
    )

    @field_validator("geometry")
    @classmethod
    def _ensure_geojson_dict(
        cls, value: dict[str, Any] | None
    ) -> dict[str, Any] | None:
        """Only accept dictionaries for GeoJSON payloads."""

        if value is None:
            return None
        if not isinstance(value, dict):
            raise TypeError("geometry must be a GeoJSON feature dictionary")
        return value

    @field_validator("address")
    @classmethod
    def _strip_address(cls, value: str | None) -> str | None:
        """Normalise the address string."""

        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None

    @field_validator("project_type")
    @classmethod
    def _strip_project_type(cls, value: str | None) -> str | None:
        """Normalise the project type string."""

        if value is None:
            return None
        cleaned = value.strip()
        return cleaned or None

    @field_validator("efficiency_ratio")
    @classmethod
    def _limit_efficiency_ratio(cls, value: float | None) -> float | None:
        """Ensure the efficiency ratio remains within a sensible range."""

        if value is None:
            return None
        return max(0.0, min(value, 1.0)) or None

    @model_validator(mode="after")
    def _require_location(cls, instance: BuildableRequest) -> BuildableRequest:
        """Ensure an address or geometry payload is supplied."""

        if not instance.address and not instance.geometry:
            raise ValueError("Either address or geometry must be provided")
        return instance


class BuildableMetrics(BaseModel):
    """Calculated buildable metrics for the site."""

    gfa_cap_m2: int
    floors_max: int
    footprint_m2: int
    nsa_est_m2: int


class BuildableRuleProvenance(BaseModel):
    """Provenance information for a rule."""

    rule_id: int
    clause_ref: str | None = None
    document_id: int | None = None
    pages: list[int] | None = None
    seed_tag: str | None = None


class BuildableRule(BaseModel):
    """Rule entry surfaced in buildable screening."""

    id: int
    authority: str
    parameter_key: str
    operator: str
    value: str
    unit: str | None = None
    provenance: BuildableRuleProvenance


class ZoneSource(BaseModel):
    """Metadata describing the source of the zoning information."""

    kind: Literal["parcel", "geometry", "unknown"]
    layer_name: str | None = None
    jurisdiction: str | None = None
    parcel_ref: str | None = None
    parcel_source: str | None = None
    note: str | None = None


class BuildableCalculation(BaseModel):
    """Intermediate calculation payload returned by the service."""

    metrics: BuildableMetrics
    zone_source: ZoneSource
    rules: list[BuildableRule]

    @computed_field  # type: ignore[prop-decorator]
    @property
    def gfa_total(self) -> float:
        """Total gross floor area permitted by the calculation."""
        return float(self.metrics.gfa_cap_m2)


class BuildableResponse(BaseModel):
    """Response payload returned by the buildable screening endpoint."""

    model_config = ConfigDict(
        json_schema_extra={"example": _BUILDABLE_RESPONSE_EXAMPLE}
    )

    input_kind: Literal["address", "geometry"]
    zone_code: str | None = None
    overlays: list[str]
    advisory_hints: list[str]
    metrics: BuildableMetrics
    zone_source: ZoneSource
    rules: list[BuildableRule]


__all__ = [
    "BUILDABLE_REQUEST_EXAMPLE",
    "BUILDABLE_RESPONSE_EXAMPLE",
    "BuildableCalculation",
    "BuildableDefaults",
    "BuildableMetrics",
    "BuildableRequest",
    "BuildableResponse",
    "BuildableRule",
    "BuildableRuleProvenance",
    "ZoneSource",
]
